File: jscomp/others/belt_MutableStack.ml
(* Copyright (C) 2017 Hongbo Zhang, Authors of ReScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)


type  'a t = { mutable root : 'a opt_cell}
and 'a opt_cell = 'a cell option
and 'a cell = {
  head : 'a ; 
  tail : 'a opt_cell
} 


let make () = {root = None}

let clear s =  s.root <- None

let copy (s : _ t) : _ t = {root = s.root}

let push s x = 
  s.root <- Some {head = x; tail = s.root}

let topUndefined (s : 'a t) = 
   match s.root with  
   | None -> Js.undefined
   | Some x -> Js.Undefined.return x.head 

let top s = 
  match s.root with 
  | None -> None 
  | Some x -> Some x.head

let isEmpty s = s.root = None

let popUndefined s =   
  match s.root with 
  | None -> Js.undefined
  | Some x -> 
    s.root <- x.tail;    
    Js.Undefined.return x.head

let pop s =     
    match s.root with 
  | None -> None
  | Some x -> 
    s.root <- x.tail;
    Some x.head



let rec lengthAux (x : _ cell) acc = 
  match x.tail with 
  | None -> acc + 1 
  | Some x -> lengthAux x (acc + 1)

let size s =   
  match s.root with 
  | None -> 0 
  | Some x -> lengthAux x 0

let rec iterAux (s : _ opt_cell) f =  
  match s with 
  | None -> ()
  | Some x -> 
    f x.head [@bs];
    iterAux x.tail f 

let forEachU s f =   
  iterAux s.root f 

let forEach s f = forEachU s (fun [@bs] x -> f x)


let rec dynamicPopIterU s  f = 
  match s.root with 
  | Some {tail; head }-> 
    s.root <- tail;
    f head [@bs] ;
    dynamicPopIterU s  f (* using root, `f` may change it*)
 | None -> ()   



let dynamicPopIter s f = dynamicPopIterU s (fun [@bs] x -> f x)

File: lib/js/belt_MutableStack.js
'use strict';

var Curry = require("./curry.js");
var Caml_option = require("./caml_option.js");

function make(param) {
  return {
          root: undefined
        };
}

function clear(s) {
  s.root = undefined;
}

function copy(s) {
  return {
          root: s.root
        };
}

function push(s, x) {
  s.root = {
    head: x,
    tail: s.root
  };
}

function topUndefined(s) {
  var x = s.root;
  if (x !== undefined) {
    return x.head;
  }
  
}

function top(s) {
  var x = s.root;
  if (x !== undefined) {
    return Caml_option.some(x.head);
  }
  
}

function isEmpty(s) {
  return s.root === undefined;
}

function popUndefined(s) {
  var x = s.root;
  if (x !== undefined) {
    s.root = x.tail;
    return x.head;
  }
  
}

function pop(s) {
  var x = s.root;
  if (x !== undefined) {
    s.root = x.tail;
    return Caml_option.some(x.head);
  }
  
}

function size(s) {
  var x = s.root;
  if (x !== undefined) {
    var _x = x;
    var _acc = 0;
    while(true) {
      var acc = _acc;
      var x$1 = _x;
      var x$2 = x$1.tail;
      if (x$2 === undefined) {
        return acc + 1 | 0;
      }
      _acc = acc + 1 | 0;
      _x = x$2;
      continue ;
    };
  } else {
    return 0;
  }
}

function iterAux(_s, f) {
  while(true) {
    var s = _s;
    if (s === undefined) {
      return ;
    }
    f(s.head);
    _s = s.tail;
    continue ;
  };
}

function forEachU(s, f) {
  iterAux(s.root, f);
}

function forEach(s, f) {
  forEachU(s, (function (x) {
          Curry._1(f, x);
        }));
}

function dynamicPopIterU(s, f) {
  while(true) {
    var match = s.root;
    if (match === undefined) {
      return ;
    }
    s.root = match.tail;
    f(match.head);
    continue ;
  };
}

function dynamicPopIter(s, f) {
  dynamicPopIterU(s, (function (x) {
          Curry._1(f, x);
        }));
}

exports.make = make;
exports.clear = clear;
exports.copy = copy;
exports.push = push;
exports.popUndefined = popUndefined;
exports.pop = pop;
exports.topUndefined = topUndefined;
exports.top = top;
exports.isEmpty = isEmpty;
exports.size = size;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.dynamicPopIterU = dynamicPopIterU;
exports.dynamicPopIter = dynamicPopIter;
/* No side effect */
